### ISSUE TITLE
nan gradient for ComplexDense with angle-based loss

### ISSUE THREAD

**Issue by jonasdaugalas** [NEGU93/cvnn#12]
@NEGU93 Thanks for this great piece of work!

I am having an issue with angle-based loss functions. See here a small example:
https://colab.research.google.com/drive/10y2eBxHMq5HCbHqOsrfKzvrJ7AV_RegZ?usp=sharing

1. instantiate a model with a single complex dense layer with one unit and no activation;
2. make an input tensor with a single "zero" sample;
3. define a reconstruction loss comparing angles of input vs output;
4. pass the input through the model, get 0 loss, and `nan` gradients.

The gradients look OK (not `nan`) when the input is not zero. Maybe I am doing something wrong?

Thanks.

**Comment by NEGU93** [comment]
Dear @jonasdaugalas ,

Thank you for your interest in my library.
What you are experiencing is indeed very puzzling.

I tried many things to know what was happening.
One interesting thing to note is that when changing `tf.math.angle` of your loss to, for example, `tf.math.abs` the error stops.
This seems to point out it's actually an incompatibility between the tf angle function and the gradient. 
Would it be a tf bug? 

[Here](https://colab.research.google.com/drive/1EABU8xXjHhnTK8JJe4qNDyYnHVxlATbP?usp=sharing) the changes I did.

**Comment by NEGU93** [comment]
So now I replicated your problem but using solely tensorflow's library in [here](https://colab.research.google.com/drive/1SkomH1STPtHX6GfB8RBDRuu-pK5VkTyk?usp=sharing). I will post this as an issue on Tensorflow

**Comment by NEGU93** [comment]
Ok, so after talking for a while on tensorflow's forum we discovered the issue.
O complex number of 0+i0 has a Nan angle, as it is exactly in the middle. Therefore, using `tf.math.angle` on it will return Nan and after that, it will crash.

Apparently, the loss of Nan-Nan = 0 (which is a weird behavior).

**Comment by NEGU93** [comment]
So I don't fully understand it yet but I have that:
1. The problem is only with `0+i0` values and no other from the complex domain.
2. To cite a [response](https://github.com/tensorflow/tensorflow/issues/50854#issuecomment-885581403) in Tensorflow forum:

> What I can gather is that nan numpy values for gradients are acceptable when those gradients are differentials of zero values, and this shouldn't cause an issue when actual complex numbers are inserted.

Although I do not yet understand why.

**Comment by NEGU93** [comment]
I will leave this issue open but according to the discussion on tensorflow I believe it's an issue the have with tf.math.angle and the gradients. I encourage you to participate on the discussion or create a new issue. From my part I am still trying to understand the problem.

**Comment by NEGU93** [comment]
[This](https://www.tensorflow.org/guide/autodiff#zeros_instead_of_none) might be helpful to solve this issue.

**Comment by ooshyun** [comment]
I got a same problem in model. Thanks for answering this question.

However, how's about if when we get the complex number 0+0j, then add the smallest number in imaginray part inducing the differentiate to 0? As previous discussion, if complex number is a+bj, then the differentiate equation atan(b/a) goes to 0. This seems like avoiding log / divide calculation to inf. Or can I ask about solutions for avoiding the differentiate tf.math.angle from nan?
